src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs

```diff
@@ -174,42 +227,36 @@ private BoundSequence VisitIndexOrRangePatternIndexerAccess(BoundIndexOrRangePat
                 _compilation.GetWellKnownType(WellKnownType.System_Index),
                 TypeCompareKind.ConsiderEverything))
             {
-                return VisitIndexPatternIndexerAccess(
-                    node.Syntax,
-                    node.Receiver,
-                    node.LengthOrCountProperty,
-                    (PropertySymbol)node.PatternSymbol,
-                    node.Argument,
-                    isLeftOfAssignment: isLeftOfAssignment);
+                return VisitIndexPatternIndexerAccess(node, isLeftOfAssignment: isLeftOfAssignment);
             }
             else
             {
                 Debug.Assert(TypeSymbol.Equals(
                     node.Argument.Type,
                     _compilation.GetWellKnownType(WellKnownType.System_Range),
                     TypeCompareKind.ConsiderEverything));
-                return VisitRangePatternIndexerAccess(
-                    node.Receiver,
-                    node.LengthOrCountProperty,
-                    (MethodSymbol)node.PatternSymbol,
-                    node.Argument);
+                Debug.Assert(!isLeftOfAssignment || node.IndexerAccess.GetRefKind() == RefKind.Ref);
+
+                return VisitRangePatternIndexerAccess(node);
             }
         }
 
-
-        private BoundSequence VisitIndexPatternIndexerAccess(
-            SyntaxNode syntax,
-            BoundExpression receiver,
-            PropertySymbol lengthOrCountProperty,
-            PropertySymbol intIndexer,
-            BoundExpression argument,
-            bool isLeftOfAssignment)
+        private BoundSequence VisitIndexPatternIndexerAccess(BoundIndexOrRangePatternIndexerAccess node, bool isLeftOfAssignment)
         {
+            Debug.Assert(node.ArgumentPlaceholders.Length == 1);
+            Debug.Assert(node.IndexerAccess is BoundIndexerAccess);
+
+            Debug.Assert(TypeSymbol.Equals(
+                node.Argument.Type,
+                _compilation.GetWellKnownType(WellKnownType.System_Index),
+                TypeCompareKind.ConsiderEverything));
+
             var F = _factory;
 
             var locals = ArrayBuilder<LocalSymbol>.GetInstance(2);
             var sideeffects = ArrayBuilder<BoundExpression>.GetInstance(2);
 
+            var receiver = node.GetReceiver();
             Debug.Assert(receiver.Type is { });
             var receiverLocal = F.StoreToTemp(
```

Comment:
>StoreToTemp [](http://example.com/codeflow?start=34&length=11)

It feels like it will be beneficial to avoid capturing temps created by the DAG rewrite.